### ISSUE TITLE
Extract image alternate text from keydef linktext

### ISSUE THREAD

**Issue by raducoravu** [dita-ot/dita-ot#2752]
According to this email exchange on the [DITA TC list](https://lists.oasis-open.org/archives/dita-comment/201707/msg00001.html), according to [DITA 1.3 spec](http://docs.oasis-open.org/dita/dita/v1.3/errata01/os/complete/part1-base/archSpec/base/processing-keyref-for-text.html):

> For empty `<image>` elements, effective content is used as alternate text, equivalent to creating an `<alt>` sub-element to hold that content.

If in my DITA Map I have a key definition:

```xml
<keydef keys="myImage" href="images/Chrysanthemums.jpg" format="jpg">
  <topicmeta>
    <linktext>Alternate text</linktext>
  </topicmeta>
</keydef>
```

and in a DITA topic I have a keyref to it:
```xml
<image keyref="myImage" id="image_tq4_4lc_p1b"/>
```
the image `<alt>` element should be computed automatically from the `<linktext>` set on the keydef `<topicmeta>`.

**Comment by robander** [comment]
I wondered about this when working on #2719 but knew I wouldn't have time to add it in. I think the final refactoring @jelovirt did before merging that item should actually make fixing this easier, but I haven't tried yet. Would be nice to get it into 2.5.2 if we can.

**Comment by robander** [comment]
Looking into this one, the current code inserts the alternate text directly inside `<image>` when the image does not already have alternate text. Of course that's not valid, but it's what I'd expect from the current code. It doesn't show up anywhere because X/HTML and PDF all explicitly look for `<alt>` and process that.

Gets a bit trickier here because technically, the deprecated `@alt` on images should probably also override anything specified in the key definition -- currently every test for "link text" is based on element content, not attribute content.

**Comment by robander** [comment]
This one was bugging me so I kept thinking about it. My eventual logic may be partly related to the fact that all the keyref link text processing is in the `endElement` method so no easy access to test if `@alt` was specified. But ignoring that..

1. The `@alt` attribute was deprecated in favor of `<alt>` way back in DITA 1.1 (2007).
1. Keys were not even defined until DITA 1.2.
1. The key resolution here, that explicitly leads to `<alt>`, wasn't fully clarified until DITA 1.3 (2015).
1. Now, 10 years after it was deprecated, I don't think we should go out of our way to support `@alt` as a way to override the key reference feature that was added more recently.

With that in mind -- @jelovirt or @raducoravu any thoughts on this commit as a fix for the current bug? 
https://github.com/robander/dita-ot/commit/80468032e8f0e961ce4bc835f2b6e94de9c72772

It currently breaks the integration tests because we've got a sample in there that changes, but I could fix that in a pull request.

**Comment by jelovirt** [comment]
I'm ok with that fix.

**Comment by raducoravu** [comment]
👍 There is a little bit of duplicate code between the two "writeAlt" methods, maybe a "writeAlt" which could received two parameters, the content as Element and the content as String, and depending on which one is not null, output that one...
One question, how about if there is an element derived from DITA image? Maybe it no longer has an alt element or maybe its "alt" element is not called "alt" anymore... Should we treat in the fix only DITA "image" elements (having the tag name "image") instead of matching the class attribute value?

**Comment by robander** [comment]
For the duplicate code -- I've just used the same pattern we already have for `writeLinktext`, so I don't really want to create a new pattern here. Can change both later if we decide it's better to do as one method.

For specializations - we should not worry about that just for this case. The same theoretical issue comes up in almost any DITA processing that changes markup, so we shouldn't do something special here. For example - almost any DITA keyword/term/link/xref element could be specialized with `EMPTY` content model, but we still insert link text for those as well. Similarly, you could specialize a topic that doesn't allow `<related-links>` but we'll still generate map based linking. None of these break anything because we are not enforcing schema validity at this point, and if special processing exists for these conditions they'll just ignore the generated content.

**Comment by raducoravu** [comment]
@robander Ok.

**Comment by infotexture** [comment]
Fixed in #2814.
